test/libsolidity/ASTPropertyTest.cpp

```diff
@@ -0,0 +1,202 @@
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+// SPDX-License-Identifier: GPL-3.0
+
+#include <test/libsolidity/ASTPropertyTest.h>
+#include <test/Common.h>
+
+#include <libsolidity/ast/ASTJsonExporter.h>
+#include <libsolidity/interface/CompilerStack.h>
+
+#include <liblangutil/Common.h>
+
+#include <boost/algorithm/string.hpp>
+#include <boost/throw_exception.hpp>
+
+#include <range/v3/algorithm/find_if.hpp>
+#include <range/v3/range/conversion.hpp>
+#include <range/v3/view/split.hpp>
+
+#include <queue>
+
+using namespace solidity::langutil;
+using namespace solidity::frontend;
+using namespace solidity::frontend::test;
+using namespace solidity;
+using namespace std;
+
+ASTPropertyTest::ASTPropertyTest(string const& _filename):
+	TestCase(_filename)
+{
+	if (!boost::algorithm::ends_with(_filename, ".sol"))
+		BOOST_THROW_EXCEPTION(runtime_error("Not a Solidity file: \"" + _filename + "\"."));
+
+	m_source = m_reader.source();
+	readExpectations();
+}
+
+void ASTPropertyTest::generateTestCaseValues(string& _values, bool _obtained)
+{
+	_values.clear();
+	for (string const& testId: m_expectationsSequence)
+	{
+		soltestAssert(m_testCases.count(testId) > 0);
+		_values +=
+			testId +
+			": " +
+			(_obtained ? m_testCases[testId].obtainedValue : m_testCases[testId].expectedValue)
+			+ "\n";
+	}
+}
+
+vector<StringPair> ASTPropertyTest::readKeyValuePairs(string const& _input)
+{
+	vector<StringPair> result;
+	for (string const& line: _input | ranges::views::split('\n') | ranges::to<vector<string>>)
+	{
+		if (line.empty())
+			continue;
+
+		vector<string> pair;
+		for (string const& element: line | ranges::views::split(':') | ranges::to<vector<string>>)
+		{
+			soltestAssert(ranges::all_of(boost::trim_copy(element), [](char c) { return isprint(c); }));
+			pair.emplace_back(boost::trim_copy(element));
+		}
+
+		soltestAssert(pair.size() == 2, "Invalid property test: " + line);
+		result.emplace_back(pair[0], pair[1]);
+	}
+	return result;
+}
+
+void ASTPropertyTest::readExpectations()
+{
+	for (auto const& [testId, testExpectation]: readKeyValuePairs(m_reader.simpleExpectations()))
+	{
+		soltestAssert(m_testCases.count(testId) == 0, "More than one expectation for test \"" + testId + "\"");
+		m_testCases.emplace(testId, ASTPropertyTestCase{testId, "", testExpectation, ""});
+		m_expectationsSequence.push_back(testId);
+	}
+	generateTestCaseValues(m_expectation, false);
+}
+
+optional<Json::Value> ASTPropertyTest::findNode(Json::Value const& _root, string_view const& _property)
+{
+	if (_root.isObject() && !_property.empty())
+	{
+		string subNode = string(_property.substr(0, _property.find_first_of('.')));
+		if (subNode != _property)
+			return findNode(_root[subNode], _property.substr(_property.find_first_of('.') + 1));
+		if (_root.isMember(subNode))
+			return make_optional(_root[subNode]);
+	}
+	return {};
+}
```

Comment:
Is this really correct? I see that when `subNode != _property` you use `_root[subNode]` without checking if it's present.

I'd also adjust naming and restructure it a bit. How about this instead?

```suggestion
optional<Json::Value> ASTPropertyTest::jsonValueByPath(Json::Value const& _node, string_view _jsonPath)
{
	if (!_node.isObject() || _jsonPath.empty())
		return {};

	string_view memberName = _jsonPath.substr(0, _jsonPath.find_first_of('.'));
	if (!_node.isMember(memberName))
		return {};
	if (memberName == _jsonPath)
		return _node[memberName];

	return findNode(_node[memberName], _jsonPath.substr(memberName.size() + 1));
}
```

Also, this seems like a pretty generic helper that might also be helpful outside of tests. Why not put it in `libsolutil/JSON.h`?

Reply:
> Is this really correct? I see that when subNode != _property you use _root[subNode] without checking if it's present.

In fact, it is not. If `subNode` is not present, operator `[]` returns a `null` value, which then I assumed that `isObject` would return `false`, however seems that's not the [case](https://jsoncpp.sourceforge.net/json__value_8cpp_source.html#l01431).
I will refactor as you suggested. 

> Also, this seems like a pretty generic helper that might also be helpful outside of tests. Why not put it in libsolutil/JSON.h?

Yeah, true. That's a good idea. I even searched for something like this before implementing it.